src/util/http/websocket/QueryHub.cpp

```diff
@@ -39,38 +34,37 @@ QueryHub::createOrAcquireDistributorInternal(QueryId queryId, bool isProvider) {
 
   auto pointerHolder =
       std::make_shared<std::optional<PointerType>>(std::nullopt);
-  auto cleanupAction = [](auto& distributors, auto queryId,
-                          auto referencePointer) -> net::awaitable<void> {
+  auto cleanupAction = [this, queryId, pointerHolder]() {
+    AD_CORRECTNESS_CHECK(pointerHolder->has_value());
+    auto referencePointer = std::move(pointerHolder->value());
     // Only erase object if we created it, otherwise we have a race condition
-    if (distributors.contains(queryId) &&
-        equals(referencePointer, distributors.at(queryId))) {
-      distributors.erase(queryId);
+    if (socketDistributors_.contains(queryId) &&
+        equals(referencePointer, socketDistributors_.at(queryId))) {
+      socketDistributors_.erase(queryId);
     }
-    // Make sure this is treated as coroutine
-    co_return;
   };
   auto distributor = std::make_shared<QueryToSocketDistributor>(
-      ioContext_, [this, queryId, coroutine = std::move(cleanupAction),
-                   pointerHolder]() mutable {
-        AD_CORRECTNESS_CHECK(pointerHolder->has_value());
-        net::co_spawn(globalStrand_,
-                      coroutine(socketDistributors_, std::move(queryId),
-                                std::move(pointerHolder->value())),
-                      net::detached);
+      ioContext_, [this, cleanupAction = std::move(cleanupAction)]() mutable {
+        net::dispatch(globalStrand_, std::move(cleanupAction));
       });
   *pointerHolder = distributor;
   socketDistributors_.emplace(queryId, distributor);
-  if (isProvider) {
-    co_await sameExecutor(distributor->signalStartIfNotStarted());
-  }
   co_return distributor;
 }
 
 // _____________________________________________________________________________
 
 net::awaitable<std::shared_ptr<QueryToSocketDistributor>>
-QueryHub::createOrAcquireDistributor(QueryId queryId, bool isProvider) {
-  return sameExecutor(
-      createOrAcquireDistributorInternal(std::move(queryId), isProvider));
+QueryHub::createOrAcquireDistributorForSending(QueryId queryId) {
+  return sameExecutor(createOrAcquireDistributorInternal(std::move(queryId)));
+}
+
+// _____________________________________________________________________________
+
+net::awaitable<std::shared_ptr<const QueryToSocketDistributor>>
+QueryHub::createOrAcquireDistributorForReceiving(QueryId queryId) {
+  auto result = co_await sameExecutor(
+      createOrAcquireDistributorInternal(std::move(queryId)));
+  co_return result;
```

Comment:
Is this now the semantics of "we might end up with multiple senders/providers for the same queryId, but this is not such a big problem because multiple calls to `signalEnd` are a non-issue because you also check the pointer-equality? Then I understand it now. (The only reason is, that messages might be lost, because the listeners are already unregistered. 

I still think you can keep track of one bool for each queryId on whether we already have picked up the listener, and reintroduce the `isSender` argument to your `...Internal` function.

Reply:
Multiple calls to `signalEnd()` will raise an assertion error, so that should be clear enough